graphjet-core/src/main/java/com/twitter/graphjet/hashing/SmallArrayBasedLongToDoubleMap.java

```diff
@@ -22,35 +22,88 @@
 import it.unimi.dsi.fastutil.ints.IntComparator;
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
 import it.unimi.dsi.fastutil.longs.LongSet;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import it.unimi.dsi.fastutil.objects.ObjectSet;
 
 /**
  * This class provides a map from long to double. It uses three primitive arrays to store long keys,
  * double values and long metadata. It only offers apis to get all keys or values or metadata at
  * once, and it does not support getting a specific key/value pair. The main purpose of this
  * implementation is to store a very small number of pairs. When a key/value pair is inserted into
  * the map, it scans through the keys array linearly and makes a decision whether to append the pair
- * or not. When the size of the map is equal to ADD_KEYS_TO_SET_THRESHOLD, it adds all keys to a set
- * and starts to use the set for dedupping.
+ * or not. When the size of the map is equal to ADD_KEYS_TO_SET_THRESHOLD, it adds all keys and
+ * metadata to a map and starts to use the map for dedupping.
  */
 public class SmallArrayBasedLongToDoubleMap {
+
+  /**
+   * Holder class of a pair of key and metadata, which is used during dedupping when the size of the
+   * map grows beyond ADD_KEYS_TO_SET_THRESHOLD.
+   */
+  private static final class Pair {
```

Comment:
All 3rd party apache pair classes are based on generics, and when it holds a pair of longs, it will autobox them into object Longs, which this internal class avoids.